src/main/java/org/elasticsearch/index/translog/fs/FsTranslogSnapshot.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.translog.fs;
+
+import org.apache.lucene.util.IOUtils;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.index.translog.TruncatedTranslogException;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class FsTranslogSnapshot implements Translog.Snapshot {
+
+    private final List<FsChannelSnapshot> orderedTranslogs;
+    private final ESLogger logger;
+    private final ByteBuffer cacheBuffer;
+    private AtomicBoolean closed = new AtomicBoolean(false);
+    private final int estimatedTotalOperations;
+
+    /**
+     * Create a snapshot of translog file channel. The length parameter should be consistent with totalOperations and point
+     * at the end of the last operation in this snapshot.
+     */
+    public FsTranslogSnapshot(List<FsChannelSnapshot> orderedTranslogs, ESLogger logger) {
+        this.orderedTranslogs = orderedTranslogs;
+        this.logger = logger;
+        int ops = 0;
+        for (FsChannelSnapshot translog : orderedTranslogs) {
+
+            final int tops = translog.estimatedTotalOperations();
+            if (tops < 0) {
+                ops = FsChannelReader.UNKNOWN;
+                break;
```

Comment:
I wanted to give a clear signal that it is unknown. this only happen when we use files from the initial recovery. Afterwards all is known.